### PR TITLE
fix: remove new lines from the end of contextual text blocks

### LINKED ISSUE #30: Refactor: make helper functions more universal/modular
This issue is a continuation of the discussion from #28 

## Problem overview

All the highlights are accompanied by a so-called representative text that acts as a full version of the sentence from which the highlight is taken. Depending on a book structure and highlighted fragments, the representative text may contain excessive symbols, like trailing spaces, newlines, tabs, or a combination of them. This results in newlines within imported highlight blocks, that add confusion as it may seem like two separate highlight blocks. For example:

```
- 📖 Chapter:: N/A
- 🔖 Context:: əˈtämik
an extremely small amount of a thing; the single irreducible unit of a larger system.

- 🎯 Highlight:: əˈtämik
an extremely small amount of a thing; the single irreducible unit of a larger system.
- 📝 Note:: Link to atomic notes
```

The initial solution was proposed in #28 and it acts as a good starting point. However, the solution doesn't cover following cases (yet):

- Trailing spaces - ` along the way. `
- Newlines and tabs
  - `reboot?\n\n`
  - `project.\n\n\n`
  - `simple answers.\n\t\t`
  - `success.\n\t\t\t`
  - `instead.\n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t`
  - `a book\n\t\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\t\t`

The proposed solution is similar to the function created for preserving indentation in text blocks: check `preserveNewlineIndentation` function in the [src/methods/aggregateDetails.ts](https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/blob/cb69361a4311a69ceba241294f4de9dcb4560d8e/src/methods/aggregateDetails.ts#L52).

## Proposal

Below are options under discussion:

1. Make the new helper more universal/modular to cover all the above mentioned cases. Modularity will keep the helper easy maintain and test, but may result in reduced code readability because if a text block requires multiple modifications, there will be a chain of helper functions performing these modifications.
2. Make a god-like helper function that takes all the helper functions and a text block as parameters and performs all the required modifications. As a result, there will be only one helper function to call but it may be harder to maintain and test this function with multiple "side effects".
3. More elegant solution to take the best from both worlds ?

### PR DESCRIPTION
**Pull request by absorpheus**
**Summary**

Fixes #30 

* Created a utility helper to remove newlines from the end of `contextualText` text blocks
* Added tests

This removes newline characters which may exist at the end of `contextualText` text blocks which are creating unnecessary new lines in Obsidian's editing mode.

This screenshot illustrates the bug in `contextualText` text blocks using the default template:

contextualText-newline-characters

This is after the fix is applied:

contextualText-after-fix

### CODE REVIEW

**Comment by bandantonio** [comment]
Thank you.
I was keeping this issue in mind for a while, and your PR is a good starting point.
In my highlights, I also found the following cases (depends on a book):

- Trailing spaces - ` along the way. `
- Newlines and tabs
  - `reboot?\n\n`
  - `project.\n\n\n`
  - `simple answers.\n\t\t`
  - `success.\n\t\t\t`
  - `instead.\n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t`
  - `a book\n\t\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\t\t`

So I thought about either making this helper more universal, or merging it (if I/we could find an elegant way) with the existing one that works with indentation: [src/methods/aggregateDetails.ts#L51](https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/blob/master/src/methods/aggregateDetails.ts#L51). That's why the issue is still here 😄

**Comment by absorpheus** [comment]
Thank you for sharing, I wasn't aware that other cases existed! I'm thinking it might be better to keep these helper functions modular so we can easily test them, however it may become cumbersome to call multiple functions nested within each other. 

We could have a 'parent' helper function that takes in an array of helper functions and the text as parameters and loops through the helper functions array, invoking the helper function with the outputted result from previous returned text each time.

I'll have a think about it. It would be helpful if we could write some tests so we can better understand the edge cases we are attempting to solve.

**Comment by absorpheus** [comment]
> * Trailing spaces - `along the way.`
> * Newlines and tabs
>   
>   * `reboot?\n\n`
>   * `project.\n\n\n`
>   * `simple answers.\n\t\t`
>   * `success.\n\t\t\t`
>   * `instead.\n\t\t\t\t\n\t\t\t\t\n\t\t\t\t\t`
>   * `a book\n\t\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\n\t\t\t\t\t\t\n\t\t\t\t\t\t\t`

Are these trailing spaces, newlines and tabs appearing only at the end of the highlighted text block (contextualText) or do they appear in other places too? e.g. in between highlighted paragraphs

**Comment by bandantonio** [comment]
> Thank you for sharing, I wasn't aware that other cases existed!

No blame at all! I should have created an issue for that.

UPD: Created #30 Let's continue this discussion there and place this PR on hold for some time.

> I'm thinking it might be better to keep these helper functions modular so we can easily test them, however it may become cumbersome to call multiple functions nested within each other.

Yes, that's how I thought initially. Modularity is a good approach both from coding and testing perspective, but eventually makes the code base a bit noisy since you end up having several helper functions doing relatively similar actions. Plus, wrapping text fragments into more than two helpers may worsen code readability.

> We could have a 'parent' helper function that takes in an array of helper functions and the text as parameters and loops through the helper functions array, invoking the helper function with the outputted result from previous returned text each time.

Yeah, that's one of the possible ways.

> I'll have a think about it. It would be helpful if we could write some tests so we can better understand the edge cases we are attempting to solve.

Yes, TDD definitely makes sense.

**Review by bandantonio — CHANGES_REQUESTED** [review]
Thank you for the updates.
I left one comment regarding tests, please check it.

Also, please make sure to do rebase of all your commits into one.
Thank you.

File: test/preserveNewlineIndentation.spec.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { describe, expect, test } from 'vitest'
+import { preserveNewlineIndentation } from 'src/utils'
+
+describe('preserveNewlineIndentation', () => {
```
[the remark this answers is not in the document]
Comment (absorpheus) [reply test/preserveNewlineIndentation.spec.ts:8]: You're welcome and done 👍🏼

Please confirm if the test meets the requirement of proper indentation when text block is wrapped. 

Also let me know if I've placed the templates and mocks in the correct files or if there's a more appropriate place where they should be put.

Thank you

[the remark this answers is not in the document]
Comment (bandantonio) [reply test/preserveNewlineIndentation.spec.ts:8]: LGTM, thank you.
You can proceed with rebase.

**Review by bandantonio — APPROVED** [review]

File: src/types.ts [hunk 2]
```diff
@@ -17,13 +17,17 @@ export interface IBookAnnotation {
     ZANNOTATIONCREATIONDATE: number;
     ZANNOTATIONMODIFICATIONDATE: number;
     ZANNOTATIONSTYLE: 0 | 1 | 2 | 3 | 4 | 5;
+    ZANNOTATIONDELETED: 0 | 1;
```
Comment (bandantonio) [inline src/types.ts:20]: This field is redundant, as the db query takes only the existing annotations
```suggestion
```

File: test/mocks/aggregateWrappedTextBlockData.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  "bookId": "THBFYNJKTGFTTVCGSAE5",
+  "bookAuthor": "Apple Inc.",
+  "bookGenre": "Technology",
+  "bookLanguage": "EN",
+  "bookLastOpenedDate": 731876693.002279,
+  "bookCoverUrl": '',
+  "annotations": [{
+    "chapter": "Another aggregated Introduction",
+    "contextualText": "This is a contextual text for the aggregated hightlight from the Apple iPhone User Guide\ncontaining a new line to test the preservation of indentation",
+    "highlight": `Chapter 1 introduces the terminology and approach\nthat we're going to use throughout this book. It examines what we actually mean by\nwords like reliability, scalability, and maintainability, and how\nwe can try to achieve these goals.`,
```
Comment (bandantonio) [inline test/mocks/aggregateWrappedTextBlockData.ts:14]: This highlight isn't consistent with the rest of the content, please update.

File: test/mocks/renderedTemplate.ts [hunk 4]
```diff
@@ -60,3 +60,13 @@ along with a new line to test the preservation of indentation
 - <small>📅 Highlight modified on:: 2024-03-11 03:04:53 PM -04:00</small>
 
 `;
+
+export const renderedWrappedTextBlockTemplateMock = `----
+
+> [!QUOTE]
+>  Chapter 1 introduces the terminology and approach
```
Comment (bandantonio) [inline test/mocks/renderedTemplate.ts:67]: Please update the test to use the proper highlight from the `aggregateWrappedTextBlockData.ts`

**Review by bandantonio — CHANGES_REQUESTED** [review]

**Comment by absorpheus** [comment]
@bandantonio I've rebased this branch with master and squashed the commits down to one commit. I force-pushed the branch so the branch on GitHub is consistent with my local branch. I think you'll need to fetch this branch again as it's history has been re-written and will be different to your local branch so apologies for the inconvenience.

Please let me know if you're happy with the changes. Thanks

**Review by bandantonio — APPROVED** [review]
